### ISSUE TITLE
Strange bug (version 5.4.1)

### ISSUE THREAD

**Issue by lucianocn** [socketio/socket.io-client-swift#311]
Hi,
Debugging with the server log I discovered that when the socket connects the emit `joinRoom` is not being sent on the first connection.

I would like to know if there is any problem in my code or if it could be something related to the lib.

```obj-c
NSURL* url = [[NSURL alloc] initWithString:SERVER_SOCKETIO_URL];
SocketIOClient *socket = [[SocketIOClient alloc] initWithSocketURL:url options:nil];

[socket on:@"connect" callback:^(NSArray* data, SocketAckEmitter* ack) {
    NSLog(@"SocketIO: Connected.");
        
    NSLog(@"SocketIO: Joining room %@.", [Authenticator getUnitID]);
    [socket emit:@"joinRoom" withItems:@[@{@"room": @"14"}]];
}];
```

Thanks,

**Comment by lucianocn** [comment]
Adding a `NSThread sleepForTimeInterval,` before the `emit`, it's working properly now.


NSURL* url = [[NSURL alloc] initWithString:SERVER_SOCKETIO_URL];
SocketIOClient *socket = [[SocketIOClient alloc] initWithSocketURL:url options:nil];

[socket on:@"connect" callback:^(NSArray* data, SocketAckEmitter* ack) {
    NSLog(@"SocketIO: Connected.");

    [NSThread sleepForTimeInterval:1.0];
    NSLog(@"SocketIO: Joining room %@.", [Authenticator getUnitID]);
    [socket emit:@"joinRoom" withItems:@[@{@"room": @"14"}]];
}];

**Comment by nuclearace** [comment]
Is the socket being retained properly? It should be a property of a class

**Comment by lucianocn** [comment]
I'm using it as:
```obj-c
@property (strong, nonatomic) SocketIOClient *socketIOClient;
```

Just the first time it is not joining the room (But if I reconnect it is working fine.)

```
2016-02-18 17:12:45.310 App[522:121799] <SocketIOClientSwift.SocketIOClient: 0x14d5c9e0>
```obj-c

**Comment by nuclearace** [comment]
If you turn logging on do you see a message about emitted when not connected?

**Comment by lucianocn** [comment]
Unsuccessful Joined Room (Log):

```obj-c
2016-02-18 22:34:31.361 Get In Manager[94874:6832969] LOG SocketIOClient: Adding handler for event: connect
2016-02-18 22:34:31.361 Get In Manager[94874:6832969] LOG SocketIOClient: Adding engine
2016-02-18 22:34:31.362 Get In Manager[94874:6832969] LOG SocketEngine: Starting engine
2016-02-18 22:34:31.363 Get In Manager[94874:6832969] LOG SocketEngine: Handshaking
2016-02-18 22:34:31.363 Get In Manager[94874:6832969] LOG SocketEnginePolling: Doing polling request
2016-02-18 22:34:32.066 Get In Manager[94874:6833167] LOG SocketEnginePolling: Got polling response
2016-02-18 22:34:32.068 Get In Manager[94874:6833148] LOG SocketEngine: Got message: 0{"sid":"XHNdScM_ZAL-XbbUAAFr","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":60000}
2016-02-18 22:34:32.070 Get In Manager[94874:6833148] LOG SocketEnginePolling: Doing polling request
2016-02-18 22:34:32.250 Get In Manager[94874:6833159] LOG SocketEnginePolling: Got polling response
2016-02-18 22:34:32.251 Get In Manager[94874:6833148] LOG SocketEngine: Got message: 40
2016-02-18 22:34:32.251 Get In Manager[94874:6833148] LOG SocketIOClient: Should parse message: 0
2016-02-18 22:34:32.252 Get In Manager[94874:6833148] LOG SocketParser: Parsing 0
2016-02-18 22:34:32.252 Get In Manager[94874:6833148] LOG SocketParser: Decoded packet as: SocketPacket {type: 0; data: []; id: -1; placeholders: 0; nsp: /}
2016-02-18 22:34:32.252 Get In Manager[94874:6833159] LOG SocketEnginePolling: Doing polling request
2016-02-18 22:34:32.253 Get In Manager[94874:6833148] LOG SocketIOClient: Socket connected
2016-02-18 22:34:32.254 Get In Manager[94874:6833148] LOG SocketIOClient: Handling event: connect with data: (
)
2016-02-18 22:34:32.317 Get In Manager[94874:6833159] LOG SocketIOClient: Emitting: 2["joinRoom",{"room":14}]
2016-02-18 22:34:32.318 Get In Manager[94874:6833159] LOG SocketEngine: Writing poll: 2["joinRoom",{"room":14}] has data: 0
2016-02-18 22:34:32.319 Get In Manager[94874:6833159] LOG SocketEnginePolling: Sending poll: 2["joinRoom",{"room":14}] as type: 4
2016-02-18 22:34:32.320 Get In Manager[94874:6833159] LOG SocketEnginePolling: Created POST string: 26:42["joinRoom",{"room":14}]
2016-02-18 22:34:32.320 Get In Manager[94874:6833159] LOG SocketEnginePolling: POSTing
2016-02-18 22:34:32.321 Get In Manager[94874:6833159] LOG SocketEnginePolling: Doing polling request
2016-02-18 22:34:32.917 Get In Manager[94874:6833151] LOG SocketEngine: Sending ws: probe as type: 2
2016-02-18 22:34:33.073 Get In Manager[94874:6833151] LOG SocketEngine: Got message: 3probe
2016-02-18 22:34:33.073 Get In Manager[94874:6833151] LOG SocketEngine: Upgrading transport to WebSockets
2016-02-18 22:34:33.074 Get In Manager[94874:6833151] LOG SocketEnginePolling: Sending poll:  as type: 6
2016-02-18 22:34:33.175 Get In Manager[94874:6833159] LOG SocketEnginePolling: Got polling response
2016-02-18 22:34:33.176 Get In Manager[94874:6833159] LOG SocketEngine: Sending ws:  as type: 5
2016-02-18 22:34:33.176 Get In Manager[94874:6833151] LOG SocketEngine: Got message: 6
2016-02-18 22:34:33.176 Get In Manager[94874:6833159] LOG SocketEngine: Flushing probe wait
```

----------------------------

Successful Joined Room (Log):

```obj-c
2016-02-18 22:37:43.694 Get In Manager[95119:6839673] LOG SocketIOClient: Adding handler for event: connect
2016-02-18 22:37:43.695 Get In Manager[95119:6839673] LOG SocketIOClient: Adding engine
2016-02-18 22:37:43.695 Get In Manager[95119:6839673] LOG SocketEngine: Starting engine
2016-02-18 22:37:43.695 Get In Manager[95119:6839673] LOG SocketEngine: Handshaking
2016-02-18 22:37:43.695 Get In Manager[95119:6839673] LOG SocketEnginePolling: Doing polling request
2016-02-18 22:37:44.389 Get In Manager[95119:6839863] LOG SocketEnginePolling: Got polling response
2016-02-18 22:37:44.390 Get In Manager[95119:6839870] LOG SocketEngine: Got message: 0{"sid":"_RC9uSfKcHL2IyXLAAFs","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":60000}
2016-02-18 22:37:44.392 Get In Manager[95119:6839870] LOG SocketEnginePolling: Doing polling request
2016-02-18 22:37:44.552 Get In Manager[95119:6839880] LOG SocketEnginePolling: Got polling response
2016-02-18 22:37:44.553 Get In Manager[95119:6839880] LOG SocketEnginePolling: Doing polling request
2016-02-18 22:37:44.553 Get In Manager[95119:6839866] LOG SocketEngine: Got message: 40
2016-02-18 22:37:44.553 Get In Manager[95119:6839866] LOG SocketIOClient: Should parse message: 0
2016-02-18 22:37:44.554 Get In Manager[95119:6839866] LOG SocketParser: Parsing 0
2016-02-18 22:37:44.554 Get In Manager[95119:6839866] LOG SocketParser: Decoded packet as: SocketPacket {type: 0; data: []; id: -1; placeholders: 0; nsp: /}
2016-02-18 22:37:44.555 Get In Manager[95119:6839866] LOG SocketIOClient: Socket connected
2016-02-18 22:37:44.555 Get In Manager[95119:6839866] LOG SocketIOClient: Handling event: connect with data: (
)
2016-02-18 22:37:44.584 Get In Manager[95119:6839877] LOG SocketIOClient: Emitting: 2["joinRoom",{"room":14}]
2016-02-18 22:37:44.584 Get In Manager[95119:6839877] LOG SocketEngine: Writing poll: 2["joinRoom",{"room":14}] has data: 0
2016-02-18 22:37:44.585 Get In Manager[95119:6839877] LOG SocketEnginePolling: Sending poll: 2["joinRoom",{"room":14}] as type: 4
2016-02-18 22:37:44.585 Get In Manager[95119:6839877] LOG SocketEnginePolling: Created POST string: 26:42["joinRoom",{"room":14}]
2016-02-18 22:37:44.586 Get In Manager[95119:6839877] LOG SocketEnginePolling: POSTing
2016-02-18 22:37:44.586 Get In Manager[95119:6839877] LOG SocketEnginePolling: Doing polling request
2016-02-18 22:37:45.262 Get In Manager[95119:6839866] LOG SocketEngine: Sending ws: probe as type: 2
2016-02-18 22:37:45.445 Get In Manager[95119:6839870] LOG SocketEngine: Got message: 3probe
2016-02-18 22:37:45.445 Get In Manager[95119:6839870] LOG SocketEngine: Upgrading transport to WebSockets
2016-02-18 22:37:45.446 Get In Manager[95119:6839870] LOG SocketEnginePolling: Sending poll:  as type: 6
2016-02-18 22:37:45.446 Get In Manager[95119:6839870] LOG SocketEnginePolling: Created POST string: 1:6
2016-02-18 22:37:45.447 Get In Manager[95119:6839870] LOG SocketEnginePolling: POSTing
2016-02-18 22:37:45.447 Get In Manager[95119:6839870] LOG SocketEnginePolling: Doing polling request
2016-02-18 22:37:45.547 Get In Manager[95119:6839876] LOG SocketEnginePolling: Got polling response
2016-02-18 22:37:45.547 Get In Manager[95119:6839876] LOG SocketEngine: Sending ws:  as type: 5
2016-02-18 22:37:45.547 Get In Manager[95119:6839870] LOG SocketEngine: Got message: 6
2016-02-18 22:37:45.548 Get In Manager[95119:6839876] LOG SocketEngine: Flushing probe wait
```

**Comment by nuclearace** [comment]
It looks like it's sending fine. What does the server say

**Comment by lucianocn** [comment]
@nuclearace I'm sure that the server is not receiving the emit, because I have one `console.log` to check it:

```javascript
var server     = require('http').createServer(),
    io         = require('socket.io')(server),
    port       = 5000;

console.log('-> SocketIO: listening on port %s', port);

io.on('connection', function (socket){
    console.log('SocketIO: Connected socket %s', socket.id);

    socket.on('joinRoom', function (data) {
        console.log('SocketIO: Joining room %s', data.room); // Not being called.
        socket.join(data.room);
    });

    socket.on('disconnect', function () {
        console.log('SocketIO: Disconnected socket %s', socket.id);
    });
});

server.listen(port);
```

**Comment by nuclearace** [comment]
I meant logging via `DEBUG=* node index.js`

**Comment by lucianocn** [comment]
I confirmed the bug, here are the logs:

When emit but server do not receive (very rare to occur):

iOS:
```obj-c
2016-02-26 13:03:01.559 App[513:36960] LOG SocketIOClient: Adding handler for event: connect
2016-02-26 13:03:01.559 App[513:36960] LOG SocketIOClient: Adding handler for event: line
2016-02-26 13:03:01.560 App[513:36960] LOG SocketIOClient: Adding engine
2016-02-26 13:03:01.562 App[513:36960] LOG SocketEngine: Starting engine
2016-02-26 13:03:01.563 App[513:36960] LOG SocketEngine: Handshaking
2016-02-26 13:03:01.563 App[513:36960] LOG SocketEnginePolling: Doing polling request
2016-02-26 13:03:02.246 App[513:36997] LOG SocketEnginePolling: Got polling response
2016-02-26 13:03:02.249 App[513:36995] LOG SocketEngine: Got message: 0{"sid":"78LKHdguTTppk-v8AABA","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":60000}
2016-02-26 13:03:02.259 App[513:36995] LOG SocketEnginePolling: Doing polling request
2016-02-26 13:03:02.449 App[513:36995] LOG SocketEnginePolling: Got polling response
2016-02-26 13:03:02.450 App[513:36995] LOG SocketEnginePolling: Doing polling request
2016-02-26 13:03:02.455 App[513:36995] LOG SocketEngine: Got message: 40
2016-02-26 13:03:02.456 App[513:36995] LOG SocketIOClient: Should parse message: 0
2016-02-26 13:03:02.458 App[513:36995] LOG SocketParser: Parsing 0
2016-02-26 13:03:02.461 App[513:36995] LOG SocketParser: Decoded packet as: SocketPacket {type: 0; data: []; id: -1; placeholders: 0; nsp: /}
2016-02-26 13:03:02.461 App[513:36995] LOG SocketIOClient: Socket connected
2016-02-26 13:03:02.461 App[513:36995] LOG SocketIOClient: Handling event: connect with data: (
)
2016-02-26 13:03:02.906 App[513:37021] LOG SocketIOClient: Emitting: 2["joinRoom",{"room":14}]
2016-02-26 13:03:02.906 App[513:37012] LOG SocketEngine: Writing poll: 2["joinRoom",{"room":14}] has data: 0
2016-02-26 13:03:02.907 App[513:37012] LOG SocketEnginePolling: Sending poll: 2["joinRoom",{"room":14}] as type: 4
2016-02-26 13:03:02.908 App[513:37012] LOG SocketEnginePolling: Created POST string: 26:42["joinRoom",{"room":14}]
2016-02-26 13:03:02.909 App[513:37012] LOG SocketEnginePolling: POSTing
2016-02-26 13:03:02.910 App[513:37012] LOG SocketEnginePolling: Doing polling request
2016-02-26 13:03:03.177 App[513:37021] LOG SocketEngine: Sending ws: probe as type: 2
2016-02-26 13:03:03.344 App[513:37021] LOG SocketEngine: Got message: 3probe
2016-02-26 13:03:03.344 App[513:37021] LOG SocketEngine: Upgrading transport to WebSockets
2016-02-26 13:03:03.345 App[513:37021] LOG SocketEnginePolling: Sending poll:  as type: 6
2016-02-26 13:03:03.441 App[513:36995] LOG SocketEnginePolling: Got polling response
2016-02-26 13:03:03.441 App[513:36995] LOG SocketEngine: Sending ws:  as type: 5
2016-02-26 13:03:03.442 App[513:36995] LOG SocketEngine: Flushing probe wait
2016-02-26 13:03:03.444 App[513:37012] LOG SocketEngine: Got message: 6
2016-02-26 13:03:27.261 App[513:37021] LOG SocketEngine: Writing ws:  has data: 0
2016-02-26 13:03:27.262 App[513:37021] LOG SocketEngine: Sending ws:  as type: 2
2016-02-26 13:03:27.479 App[513:37021] LOG SocketEngine: Got message: 3
```
-------------------

App.js
```javascript
engine intercepting request for path "/socket.io/" +11s
engine handling "GET" http request "/socket.io/?transport=polling&b64=1" +0ms
engine handshaking client "78LKHdguTTppk-v8AABA" +0ms
engine:socket sending packet "open" ({"sid":"78LKHdguTTppk-v8AABA","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":60000}) +1ms
engine:polling setting request +0ms
engine:socket flushing buffer to transport +0ms
engine:polling writing "97:0{"sid":"78LKHdguTTppk-v8AABA","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":60000}" +0ms
engine:socket executing batch send callback +0ms
socket.io:server incoming connection with id 78LKHdguTTppk-v8AABA +0ms
socket.io:client connecting to namespace / +0ms
socket.io:namespace adding socket to nsp / +2ms
socket.io:socket socket connected - writing packet +1ms
socket.io:socket joining room /#78LKHdguTTppk-v8AABA +0ms
socket.io:client writing packet {"type":0,"nsp":"/"} +7ms
socket.io-parser encoding packet {"type":0,"nsp":"/"} +0ms
socket.io-parser encoded {"type":0,"nsp":"/"} as 0 +0ms
engine:socket sending packet "message" (0) +0ms
SocketIO: Connected socket /#78LKHdguTTppk-v8AABA
socket.io:socket joined room /#78LKHdguTTppk-v8AABA +1ms
engine intercepting request for path "/socket.io/" +172ms
engine handling "GET" http request "/socket.io/?transport=polling&b64=1&sid=78LKHdguTTppk-v8AABA" +1ms
engine setting new request for existing client +0ms
engine:polling setting request +0ms
engine:socket flushing buffer to transport +0ms
engine:polling writing "2:40" +0ms
engine:socket executing batch send callback +0ms
engine intercepting request for path "/socket.io/" +188ms
engine handling "GET" http request "/socket.io/?transport=polling&b64=1&sid=78LKHdguTTppk-v8AABA" +0ms
engine setting new request for existing client +0ms
engine:polling setting request +0ms
engine upgrading existing transport +553ms
engine:socket might upgrade socket transport from "polling" to "websocket" +1ms
engine:ws received "2probe" +167ms
engine:ws writing "3probe" +0ms
engine:socket writing a noop packet to polling for fast upgrade +101ms
engine:polling writing "1:6" +0ms
engine:ws received "5" +163ms
engine:socket got upgrade packet - upgrading +0ms
engine:polling closing +0ms
engine:polling transport discarded - closing right away +0ms
engine intercepting request for path "/socket.io/" +15ms
engine handling "POST" http request "/socket.io/?transport=polling&b64=1&sid=78LKHdguTTppk-v8AABA" +0ms
engine bad request: unexpected transport without upgrade +0ms
engine:ws received "6" +187ms
engine:socket packet +0ms
```

**Comment by lucianocn** [comment]
When everything goes fine:

iOS:

```objc-c
2016-02-26 13:06:27.717 App[518:37771] LOG SocketIOClient: Adding handler for event: connect
2016-02-26 13:06:27.717 App[518:37771] LOG SocketIOClient: Adding handler for event: line
2016-02-26 13:06:27.717 App[518:37771] LOG SocketIOClient: Adding engine
2016-02-26 13:06:27.720 App[518:37771] LOG SocketEngine: Starting engine
2016-02-26 13:06:27.720 App[518:37771] LOG SocketEngine: Handshaking
2016-02-26 13:06:27.720 App[518:37771] LOG SocketEnginePolling: Doing polling request
2016-02-26 13:06:28.403 App[518:37805] LOG SocketEnginePolling: Got polling response
2016-02-26 13:06:28.407 App[518:37805] LOG SocketEngine: Got message: 0{"sid":"Y42PJAfxZ-Cc5IxiAABB","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":60000}
2016-02-26 13:06:28.415 App[518:37805] LOG SocketEnginePolling: Doing polling request
2016-02-26 13:06:28.585 App[518:37816] LOG SocketEnginePolling: Got polling response
2016-02-26 13:06:28.585 App[518:37816] LOG SocketEnginePolling: Doing polling request
2016-02-26 13:06:28.587 App[518:37816] LOG SocketEngine: Got message: 40
2016-02-26 13:06:28.587 App[518:37816] LOG SocketIOClient: Should parse message: 0
2016-02-26 13:06:28.588 App[518:37816] LOG SocketParser: Parsing 0
2016-02-26 13:06:28.590 App[518:37816] LOG SocketParser: Decoded packet as: SocketPacket {type: 0; data: []; id: -1; placeholders: 0; nsp: /}
2016-02-26 13:06:28.590 App[518:37816] LOG SocketIOClient: Socket connected
2016-02-26 13:06:28.591 App[518:37816] LOG SocketIOClient: Handling event: connect with data: (
)
2016-02-26 13:06:28.873 App[518:37816] LOG SocketIOClient: Emitting: 2["joinRoom",{"room":14}]
2016-02-26 13:06:28.875 App[518:37816] LOG SocketEngine: Writing poll: 2["joinRoom",{"room":14}] has data: 0
2016-02-26 13:06:28.876 App[518:37816] LOG SocketEnginePolling: Sending poll: 2["joinRoom",{"room":14}] as type: 4
2016-02-26 13:06:28.876 App[518:37816] LOG SocketEnginePolling: Created POST string: 26:42["joinRoom",{"room":14}]
2016-02-26 13:06:28.876 App[518:37816] LOG SocketEnginePolling: POSTing
2016-02-26 13:06:28.877 App[518:37816] LOG SocketEnginePolling: Doing polling request
2016-02-26 13:06:29.259 App[518:37804] LOG SocketEngine: Sending ws: probe as type: 2
2016-02-26 13:06:29.418 App[518:37816] LOG SocketEngine: Got message: 3probe
2016-02-26 13:06:29.418 App[518:37816] LOG SocketEngine: Upgrading transport to WebSockets
2016-02-26 13:06:29.419 App[518:37816] LOG SocketEnginePolling: Sending poll:  as type: 6
2016-02-26 13:06:29.522 App[518:37802] LOG SocketEnginePolling: Got polling response
2016-02-26 13:06:29.523 App[518:37802] LOG SocketEngine: Sending ws:  as type: 5
2016-02-26 13:06:29.525 App[518:37802] LOG SocketEngine: Flushing probe wait
2016-02-26 13:06:29.525 App[518:37816] LOG SocketEngine: Got message: 6
```

App.js:

```javascript
socket.io:socket joined room /#E1gExsEvG3szH054AABC +0ms
engine intercepting request for path "/socket.io/" +196ms
engine handling "GET" http request "/socket.io/?transport=polling&b64=1&sid=E1gExsEvG3szH054AABC" +0ms
engine setting new request for existing client +1ms
engine:polling setting request +0ms
engine:socket flushing buffer to transport +0ms
engine:polling writing "2:40" +0ms
engine:socket executing batch send callback +0ms
engine intercepting request for path "/socket.io/" +311ms
engine handling "GET" http request "/socket.io/?transport=polling&b64=1&sid=E1gExsEvG3szH054AABC" +0ms
engine setting new request for existing client +0ms
engine:polling setting request +0ms
engine upgrading existing transport +423ms
engine:socket might upgrade socket transport from "polling" to "websocket" +0ms
engine:ws received "2probe" +159ms
engine:ws writing "3probe" +0ms
engine:socket writing a noop packet to polling for fast upgrade +100ms
engine:polling writing "1:6" +0ms
engine intercepting request for path "/socket.io/" +17ms
engine handling "POST" http request "/socket.io/?transport=polling&b64=1&sid=E1gExsEvG3szH054AABC" +0ms
engine setting new request for existing client +1ms
engine:polling received "26:42["joinRoom",{"room":14}]" +0ms
engine:socket packet +0ms
socket.io-parser decoded 2["joinRoom",{"room":14}] as {"type":2,"nsp":"/","data":["joinRoom",{"room":14}]} +0ms
socket.io:socket got packet {"type":2,"nsp":"/","data":["joinRoom",{"room":14}]} +0ms
socket.io:socket emitting event ["joinRoom",{"room":14}] +0ms
socket.io:socket joining room 14 +0ms
socket.io:socket joined room 14 +1ms
engine:ws received "5" +160ms
engine:socket got upgrade packet - upgrading +0ms
engine:polling closing +0ms
engine:polling transport discarded - closing right away +0ms
engine:ws received "6" +193ms
engine:socket packet +0ms
```

**Comment by lucianocn** [comment]
PS: You can check that the iOS SocketIO log are different.

**Comment by nuclearace** [comment]
Oh I see, it's because the client is sending a post while it's upgrading. Quick fix is to force polling or websockets. I'll have to go through the upgrade code again

**Comment by lucianocn** [comment]
@nuclearace thanks for the quick answer ;)

Do you think adding `[NSThread sleepForTimeInterval:1.0];` is a secure fix?

**Comment by lucianocn** [comment]
@nuclearace any news on it?

**Comment by nuclearace** [comment]
I haven't had time to debug it

**Comment by nuclearace** [comment]
You can try the [development](https://github.com/socketio/socket.io-client-swift/tree/development) branch, I added some commits there that deal with polling and websocket

**Comment by lucianocn** [comment]
@nuclearace alright! thanks!

**Comment by nuclearace** [comment]
Does that branch fix it?

**Comment by lucianocn** [comment]
@nuclearace this bug has been fixed?

I was a little away from the project, so I jumped version 5.5.0 to 6.1.1 :P

**Comment by nuclearace** [comment]
I'm not sure, I added some checks I think, but I haven't seen it.

**Comment by tolgaytoklar** [comment]
Hey @lucianocn How did you handle this problem? 

Looks like I have same problem.

https://github.com/socketio/socket.io-client-swift/issues/385

**Comment by lucianocn** [comment]
Hi @tolgaytoklar,
Unfortunately I moved my project to Firebase ;(

**Comment by JustinLG9** [comment]
For anyone still experiencing a similar issue:

TLDR: Forcing websockets fixed my issue: ```let config = SocketIOClientConfiguration(arrayLiteral: .forceWebsockets(true))```


My initial socket connection was getting hung up and failing intermittently (about 75% of the time) but would connect successfully as soon as a retry was attempted. I got the following logs from the initial connection attempt:

```
2022-01-12T04:08:10.047227+00:00 heroku[router]: at=info method=GET path="/socket.io/?transport=polling&b64=1&EIO=4" host=redacted request_id=44e8c2fe-9c59-4a44-b417-47ec2047e61e fwd="99.3.64.131" dyno=web.1 connect=0ms service=21ms status=200 bytes=265 protocol=https
2022-01-12T04:08:10.035444+00:00 app[web.1]: 2022-01-12T04:08:10.035Z engine intercepting request for path "/socket.io/"
2022-01-12T04:08:10.036253+00:00 app[web.1]: 2022-01-12T04:08:10.036Z engine handling "GET" http request "/socket.io/?transport=polling&b64=1&EIO=4"
2022-01-12T04:08:10.047270+00:00 app[web.1]: 2022-01-12T04:08:10.047Z engine handshaking client "SCUlOmxSvmqzLtsdAAAA"
2022-01-12T04:08:10.324204+00:00 app[web.1]: 2022-01-12T04:08:10.324Z engine intercepting request for path "/socket.io/"
2022-01-12T04:08:10.324364+00:00 app[web.1]: 2022-01-12T04:08:10.324Z engine handling "GET" http request "/socket.io/?transport=polling&b64=1&EIO=4&sid=SCUlOmxSvmqzLtsdAAAA"
2022-01-12T04:08:10.324739+00:00 app[web.1]: 2022-01-12T04:08:10.324Z engine setting new request for existing client
2022-01-12T04:08:11.048153+00:00 app[web.1]: 2022-01-12T04:08:11.048Z engine upgrading existing transport
2022-01-12T04:08:11.241765+00:00 heroku[router]: at=info method=GET path="/socket.io/?transport=polling&b64=1&EIO=4&sid=SCUlOmxSvmqzLtsdAAAA" host=redacted request_id=648d58b1-faa2-4536-befc-b3d30451a675 fwd="99.3.64.131" dyno=web.1 connect=0ms service=925ms status=200 bytes=168 protocol=https
2022-01-12T04:08:11.400725+00:00 heroku[router]: at=info method=POST path="/socket.io/?transport=polling&b64=1&EIO=4&sid=SCUlOmxSvmqzLtsdAAAA" host=redacted request_id=5635b7d4-3bdd-4498-a574-c0c5cbc3e6af fwd="99.3.64.131" dyno=web.1 connect=0ms service=4ms status=400 bytes=221 protocol=https
2022-01-12T04:08:11.402416+00:00 app[web.1]: 2022-01-12T04:08:11.402Z engine intercepting request for path "/socket.io/"
2022-01-12T04:08:11.402600+00:00 app[web.1]: 2022-01-12T04:08:11.402Z engine handling "POST" http request "/socket.io/?transport=polling&b64=1&EIO=4&sid=SCUlOmxSvmqzLtsdAAAA"
2022-01-12T04:08:11.402815+00:00 app[web.1]: 2022-01-12T04:08:11.402Z engine bad request: unexpected transport without upgrade
```

The last line shows the problem: 

```engine bad request: unexpected transport without upgrade```

Something happened when trying to upgrade the socket connection from HTTP polling to websocket. Not sure what the issue was, have read about several different potential causes but was only happening on iOS, the same connection on Android worked fine.

@nuclearace suggestion of forcing websockets fixed the issue and works with my server configuration.
